core/src/test/java/com/linecorp/armeria/client/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,269 @@
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.handler.codec.http.HttpHeaderValues;
+
+public class InitiateConnectionShutdownTest {
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/ok", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
+            sb.service("/delayed_ok", (ctx, req) -> HttpResponse.delayed(HttpResponse.of(HttpStatus.OK),
+                                                                         Duration.ofMillis(500)));
+        }
+    };
+
+    private static final class CompletedResult {
+        private static final CompletedResult NOT_COMPLETED = new CompletedResult();
+        private final boolean completed;
+        @Nullable
+        private final Throwable exception;
+
+        private CompletedResult() {
+            completed = false;
+            exception = null;
+        }
+
+        private CompletedResult(@Nullable Throwable exception) {
+            completed = true;
+            this.exception = exception;
+        }
+    }
+
+    final List<RequestLogAccess> requestLogAccesses = Lists.newCopyOnWriteArrayList();
+    final AtomicReference<CompletedResult> completedResult = new AtomicReference<>(
+            CompletedResult.NOT_COMPLETED);
+
+    @BeforeEach
+    void resetVariables() {
+        requestLogAccesses.clear();
+        completedResult.set(CompletedResult.NOT_COMPLETED);
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "H1C, /ok",
+            "H1C, /delayed_ok",
+            "H2C, /ok",
+            "H2C, /delayed_ok",
+    })
+    void testConnectionShutdownCompletedExceptionallyWhenChannelNotAcquired(SessionProtocol protocol,
+                                                                            String path) throws Exception {
+        final CountingConnectionPoolListener countingListener = new CountingConnectionPoolListener();
+        try (ClientFactory clientFactory = getClientFactory(countingListener)) {
+            final WebClient client = WebClient.builder(server.uri(protocol)).factory(clientFactory).decorator(
+                    (delegate, ctx, req) -> {
+                        ClientPendingThrowableUtil.setPendingThrowable(ctx, new IllegalArgumentException());
+                        initiateConnectionShutdown(ctx);
+                        return delegate.execute(ctx, req);
+                    }).build();
+
+            try {
+                client.blocking().get(path);
+            } catch (UnprocessedRequestException ignore) {
+            }
+
+            assertNoOpenedConnectionNow(countingListener);
+            assertThat(completedResult.get().exception).isInstanceOf(UnprocessedRequestException.class);
+            assertThat(completedResult.get().exception.getCause()).isInstanceOf(IllegalArgumentException.class);
```

Comment:
nit: Should we extract `IllegalArgumentException` and compare the value with `isSameAs()`?
```java
assertThat(completedResult.get().exception.getCause())
    .isSameAs(illegalArgumentException);
```